src/main/java/picard/sam/AbstractAlignmentMerger.java

```diff
@@ -751,54 +755,105 @@ private void transferAlignmentInfoToPairedRead(final SAMRecord firstUnaligned, f
 
                 // Innies only -- do we need to do anything else about jumping libraries?
                 if (pos.getAlignmentStart() < neg.getAlignmentEnd()) {
-                    final int posDiff = pos.getAlignmentEnd() - neg.getAlignmentEnd();
-                    final int negDiff = pos.getAlignmentStart() - neg.getAlignmentStart();
-
-                    if (posDiff > 0) {
-                        final List<CigarElement> elems = new ArrayList<>(pos.getCigar().getCigarElements());
-                        Collections.reverse(elems);
-                        final int clipped = lengthOfSoftClipping(elems.iterator());
-                        final int clipFrom = pos.getReadLength() - posDiff - clipped + 1;
-                        CigarUtil.softClip3PrimeEndOfRead(pos, Math.min(pos.getReadLength(), clipFrom));
-                    }
 
-                    if (negDiff > 0) {
-                        final int clipped = lengthOfSoftClipping(neg.getCigar().getCigarElements().iterator());
-                        final int clipFrom = neg.getReadLength() - negDiff - clipped + 1;
-                        CigarUtil.softClip3PrimeEndOfRead(neg, Math.min(neg.getReadLength(), clipFrom));
+                    //need to consider unclipped positions because often the read through bases have already been soft-clipped
+                    final int posClipFrom = getReadPositionToClipFrom(pos, neg.getUnclippedEnd() + 1);
+                    final int negClipFrom = getReadPositionToClipFrom(neg, pos.getUnclippedStart() - 1);
+                    if (posClipFrom == negClipFrom && posClipFrom > 0) {
+                        final int clipFrom = posClipFrom;
+
+                        if (clippingOperator == CigarOperator.HARD_CLIP) {
+                            moveClippedBasesToTag(read1, clipFrom);
+                            moveClippedBasesToTag(read2, clipFrom);
+                        }
+
+                        CigarUtil.clip3PrimeEndOfRead(read1, clipFrom, clippingOperator);
+                        CigarUtil.clip3PrimeEndOfRead(read2, clipFrom, clippingOperator);
                     }
                 }
             }
         }
     }
 
-    /** Returns the number of soft-clipped bases until a non-soft-clipping element is encountered. */
-    private static int lengthOfSoftClipping(Iterator<CigarElement> iterator) {
-        int clipped = 0;
-        while (iterator.hasNext()) {
-            final CigarElement elem = iterator.next();
-            if (elem.getOperator() != CigarOperator.SOFT_CLIP && elem.getOperator() != CigarOperator.HARD_CLIP) {
-                break;
+    private static void moveClippedBasesToTag(final SAMRecord rec, final int clipFrom) {
+
+        final byte[] bases = rec.getReadBases();
+        final byte[] baseQualities = rec.getBaseQualities();
+        final int readLength = rec.getReadLength();
+
+        if(rec.getReadNegativeStrandFlag()) {
```

Comment:
I think that the bases should be stored in original read order, so that in order to revert, you just need to stick these on the ends of the reverted records.